Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp

```diff
@@ -19,96 +19,366 @@
+	if (!bSuccefullyCreatedReducedGrid)
 	{
-		FRandomStream NPCStream;
-		NPCStream.Initialize(FCrc::MemCrc32(&TotalNPCs, sizeof(TotalNPCs)));
-		for (int i = 0; i < TotalNPCs; i++)
+		return;
+	}
+
+	// Sort the clusters by "distance from centre of area" ascending.
+	// This ensures we iterate from closest to furthest cluster.
+	ClusterPoints.Sort([this](const FVector& Point1, const FVector& Point2)
+	{
+		const float Distance1 = FVector::Distance(Point1, WorldPosition);
+		const float Distance2 = FVector::Distance(Point2, WorldPosition);
+		return Distance2 > Distance1;
+	});
+
+	//Remove any excess areas
+	ClusterPoints.RemoveAt(MaxClusters, ClusterPoints.Num() - MaxClusters);
+
+	for (int32 i = 0; i < ClusterPoints.Num() && i < MaxClusters; ++i)
+	{
+		const FVector& ClusterPoint = ClusterPoints[i];
+		USpawnCluster* NewSpawnCluster = NewObject<USpawnCluster>(this);
+		if (NewSpawnCluster != nullptr)
 		{
-			FVector PointA = NPCStream.VRand()*RoamRadius;
-			FVector PointB = NPCStream.VRand()*RoamRadius;
-			PointA.Z = PointB.Z = 0.0f;
-			NPCRunPoints.Emplace(FBlackboardValues{ PointA, PointB });
+			NewSpawnCluster->WorldPosition = ClusterPoint;
+			NewSpawnCluster->Width = MinDistanceBetweenClusters;
+			NewSpawnCluster->Height = MinDistanceBetweenClusters;
+			NewSpawnCluster->MaxSpawnPoints = MaxSpawnPointsPerCluster;
+
+			NewSpawnCluster->GenerateSpawnPoints();
+			SpawnClusters.Add(NewSpawnCluster);
 		}
 	}
 }
 
-void ABenchmarkGymGameMode::CheckCmdLineParameters()
+TArray<AActor*> USpawnArea::CreateSpawnPointActors()
 {
-	if (bInitializedCustomSpawnParameters)
+	if (SpawnClusters.Num() == 0)
 	{
-		return;
+		return {};
 	}
 
-	ParsePassedValues();
-	StartCustomNPCSpawning();
+	TArray<AActor*> OutSpawnPointActors;
+	for (USpawnCluster* Cluster : SpawnClusters)
+	{
+		OutSpawnPointActors += Cluster->CreateSpawnPointActors();
+	}
+	return OutSpawnPointActors;
+}
 
-	bInitializedCustomSpawnParameters = true;
+// --- USpawnManager ---
+
+void USpawnManager::GenerateSpawnPoints(const int32 ZoneRows, const int32 ZoneCols, const int32 ZoneWidth, const int32 ZoneHeight,
+	const int32 ZoneClusters, const int32 BoundaryClusters,
+	const int32 MaxSpawnPointsPerCluster, const int32 MinDistanceBetweenClusters)
+{
+	GenerateSpawnAreas(ZoneRows, ZoneCols, ZoneWidth, ZoneHeight, ZoneClusters, BoundaryClusters, MaxSpawnPointsPerCluster, MinDistanceBetweenClusters);
+	CreateSpawnPointActors(ZoneClusters, BoundaryClusters, MaxSpawnPointsPerCluster);
 }
 
-void ABenchmarkGymGameMode::StartCustomNPCSpawning()
+AActor* USpawnManager::GetSpawnPointActorByIndex(const int32 Index) const
 {
-	ClearExistingSpawnPoints();
-	GenerateSpawnPointClusters(NumPlayerClusters);
+	const int32 NumSpawnPoints = GetNumSpawnPoints();
+	if (NumSpawnPoints == 0)
+	{
+		return nullptr;
+	}
+	return SpawnPointActors[Index % NumSpawnPoints];
```

Comment:
Why do we want to wrap here?

Reply:
There are only so many spawn point actors (NumSpawnPoints). We always want to return a valid spawn point actor so we looop.

Reply:
So what is the point on the index, why not just give out the next spawnpoint?

Reply:
You mean have an internal counter that tracks what the next spawn point is? If so this reduces the utility of the function, utility that we actually do use. One point in the code specifically asks each spawn area for a spawn actor at index 0 (the center).